06_puppy/alfredxiao/syncstore.go

```diff
@@ -0,0 +1,57 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+)
+
+type syncStore struct {
+	data sync.Map
+}
+
+func (s *syncStore) CreatePuppy(p Puppy) string {
+	id := uuid.New().String()
+
+	p.ID = id
+	s.data.Store(id, p)
+	return id
+}
+
+func (s *syncStore) ReadPuppy(id string) (Puppy, error) {
+	p, ok := s.data.Load(id)
+	if !ok {
+		return Puppy{}, fmt.Errorf("puppy with ID[%s] does not exists", id)
+	}
+
+	return p.(Puppy), nil
+}
+
+func (s *syncStore) UpdatePuppy(id string, p Puppy) error {
+	if id != p.ID {
+		return fmt.Errorf("bad update request, two IDs (%s, %s) do not match",
+			id, p.ID)
+	}
+
+	_, ok := s.data.Load(id)
+	if !ok {
+		return fmt.Errorf("puppy with ID[%s] does not exists", p.ID)
+	}
+
+	s.data.Store(id, p)
```

Comment:
load and store need to be guarded with [Mutex](https://golang.org/pkg/sync/#Mutex) or [RWMutex](https://golang.org/pkg/sync/#RWMutex) to avoid race conditions.
dame for delete.